### ISSUE TITLE
NVDA not reading control labels in Station Playlist Studio dialogs

### ISSUE THREAD

**Issue by neosonic2** [josephsl/stationPlaylist#160]
I've just installed the latest version of the SPL add-on into NVDA 2022.4 running on the ARM architecture version of Windows 11, along with the latest version (6.01) of Station Playlist Studio. Note that Windows 11 is running in a virtual machine through VMware Fusion on an Apple Silicon M1 Max MacBook Pro, and the virtual machine has 4 CPU cores and 16 GB RAM allocated to it.

The add-on works well, however I've noticed that controls, namely editable text fields, within some SPL Studio dialogs are not labeled properly or at all by NVDA, making it very difficult for me to know what to enter into these fields. I'm not sure if this is an issue specifically with the add-on, or with NVDA or SPL Studio itself, so I figured I'd post here first.

The best example of this issue is the Insert Tracks dialog. Pressing Alt+C to jump to the search criteria fields yields several that are unlabeled, along with others that seem to have multiple labels appended to them so I can't tell which is which. I assume the first three fields in the search criteria area are for title, artist, album, or something similar, but don't know this for sure as I do not have the eyesight required to see what's on the computer screen.

By "unlabeled," I mean that NVDA simply speaks, "edit blank" when tabbing to or entering an y of these fields.

How can this issue be rectified?

Thank you in advance for any assistance provided.

**Comment by josephsl** [comment]
Hi, this is mostly an issue with Studio itself – these edit fields are laid out on a grid with no discernable labels. I tried to resolve this several times to no avail, and with the impending end of life of the add-on, I don’t think this will get resolved anytime soon (sorry). Thanks.

**Comment by neosonic2** [comment]
Thanks for the quick response. Your comment about the impending EOL of the add-on intrigues me; is there or will there be a replacement add-on available for NVDA users? This add-on was the first thing I found when searching.

**Comment by ChrisDuffley** [comment]
If my learning of Python and NVDA source code succeeds in time, this is a project that I plan to attack and keep maintaining in the next year or so. I will send out an email to the NVDA Add-Ons and Users lists on Groups.IO if I decide to go this route. There are chances that I could rectify this via WindowControlID's but I'm not sure if that can be feasible. Thanks.

**Comment by josephsl** [comment]
Hi, I don’t think there will be a replacement. Originally I decided to end active maintenance for this add-on last year to focus on other things in life but did promise to users that I will come back to update the add-on one last time later in 2022 if no-one shows up to maintain it. A few days ago I sent out a notice to NVDA community informing folks that I’m ending maintenance for SPL and eight other add-ons for real this time mostly to keep up with graduate school. Thanks for your interest.

**Comment by neosonic2** [comment]
Would this be something that warrants raising a support request to the Station Playlist developers? Since based on the comments here it seems the issue ultimately lies with SPL itself, and also exists in various sections of the Options dialog and other places as well.

**Comment by josephsl** [comment]
Hi, I don’t think control ID’s would work as they change whenever you open insert tracks dialog (last time I checked), a thing that’s been a constant source of frustration for me throughout my time maintaining this add-on for the last nine years. SPL add-on is one of the most complex add-ons out there, and served as a way to learn more about NVDA internals until a certain add-on (shall remain nameless as the repo is different) gave me more reasons to explore NVDA source code and kept me up all night several times. Thanks.

**Comment by josephsl** [comment]
Hi, yes, please do let the developers know about this – or rather, le them know that you are also experiencing this, as the issue may lie not just with SPL but also with Delphi and whatever GUI toolkit was used to code Studio itself. Thanks.

**Comment by neosonic2** [comment]
Ah, I should have known Delphi was used to develop SPL. Before I moved to Station Playlist, I used Spacial Audio’s SAM Broadcaster Pro software, and noticed the same issue in its configuration options dialog - most edit fields were unlabeled. I’ll raise an issue with the support team about the problem. Not that this is probably any help, but JAWS for Windows seems to read the dialog correctly, but of course that’s an entirely different program and I don’t have a JAWS license as I prefer the open source (and free) nature of NVDA.

By the way, a huge thank you for all of your work on this add-on - it’s been amazing thus far and, as a result, Station Playlist Studio has been far more accessible a broadcasting program for me than SAM Broadcaster ever was in the 10 years I used it.

**Comment by neosonic2** [comment]
To follow up on this, I reached out to the Station Playlist support team and let them know about the concerns detailed in this Github issue (and even sent them a link to it). This was their response:

"We have no ideas how to resolve this for you.  The JAWS scripts developer said we should change all the graphical labels that appear right before or above every field with a Windows StaticText component, as these have a Windows Handle which screen readers can read.  We spent several days doing that within all the software about 8 years ago.  But perhaps NVDA can’t see these text labels in front of the fields for some reason.  I suspect the issue is with our development system which doesn’t support the IAccessible interface. You can raise the issue with NVDA if you wish.  If JAWS can do it then there must be a way."

**Comment by josephsl** [comment]
Hi, I suspect JAWS scripts uses display scraping to figure out what’s going on as window handles may not be reliable in this situation. I wish IAccessible/MSAA was supported by the GUI toolkit SPL uses (UI Automation might be helpful if it works in this situation). Thanks.
